### PR TITLE
Added CORE-V multiply accumulate support

### PR DESCRIPTION
**Pull request by jessicamills**
gas/ChangeLog.COREV:

	* config/tc-riscv.c: Added CORE-V multiply accumulate support.

gas/testsuite/ChangeLog.COREV:

	* gas/riscv/cv-mac-01.d: Added new test.
	* gas/riscv/cv-mac-01.l: Likewise.
	* gas/riscv/cv-mac-01.s: Likewise.
	* gas/riscv/cv-mac-02.d: Likewise.
	* gas/riscv/cv-mac-02.l: Likewise.
	* gas/riscv/cv-mac-02.s: Likewise.
	* gas/riscv/cv-mac-03.d: Likewise.
	* gas/riscv/cv-mac-03.l: Likewise.
	* gas/riscv/cv-mac-03.s: Likewise.
	* gas/riscv/cv-mac-04.d: Likewise.
	* gas/riscv/cv-mac-04.l: Likewise.
	* gas/riscv/cv-mac-04.s: Likewise.
	* gas/riscv/cv-mac-05.d: Likewise.
	* gas/riscv/cv-mac-05.l: Likewise.
	* gas/riscv/cv-mac-05.s: Likewise.
	* gas/riscv/cv-mac-06.d: Likewise.
	* gas/riscv/cv-mac-06.l: Likewise.
	* gas/riscv/cv-mac-06.s: Likewise.
	* gas/riscv/cv-mac-mac.d: Likewise.
	* gas/riscv/cv-mac-mac.s: Likewise.
	* gas/riscv/cv-mac-machhsn.d: Likewise.
	* gas/riscv/cv-mac-machhsn.s: Likewise.
	* gas/riscv/cv-mac-machhsrn.d: Likewise.
	* gas/riscv/cv-mac-machhsrn.s: Likewise.
	* gas/riscv/cv-mac-machhun.d: Likewise.
	* gas/riscv/cv-mac-machhun.s: Likeiwse.
	* gas/riscv/cv-mac-machhurn.d: Likewise.
	* gas/riscv/cv-mac-machhurn.s: Likewise.
	* gas/riscv/cv-mac-machhurn.d: Likewise.
	* gas/riscv/cv-mac-machhurn.s: Likewise.
	* gas/riscv/cv-mac-macsrn.d: Likewise.
	* gas/riscv/cv-mac-macsrn.s: Likewise.
	* gas/riscv/cv-mac-macun.d: Likewise.
	* gas/riscv/cv-mac-macun.s: Likewise.
	* gas/riscv/cv-mac-macurn.d: Likewise.
	* gas/riscv/cv-mac-macurn.s: Likewise.
	* gas/riscv/cv-mac-msu.d: Likewise.
	* gas/riscv/cv-mac-msu.s: Likewise.
	* gas/riscv/cv-mac-mulhhs.d: Likewise.
	* gas/riscv/cv-mac-mulhhs.s: Likewise.
	* gas/riscv/cv-mac-mulhhsn.d: Likewise.
	* gas/riscv/cv-mac-mulhhsn.s: Likewise.
	* gas/riscv/cv-mac-mulhhsrn.d: Likewise.
	* gas/riscv/cv-mac-mulhhsrn.s: Likewise.
	* gas/riscv/cv-mac-mulhhu.d: Likewise.
	* gas/riscv/cv-mac-mulhhu.s: Likewise.
	* gas/riscv/cv-mac-mulhhun.d: Likewise.
	* gas/riscv/cv-mac-mulhhun.s: Likewise.
	* gas/riscv/cv-mac-mulhhurn.d: Likewise.
	* gas/riscv/cv-mac-mulhhurn.s: Likewise.
	* gas/riscv/cv-mac-muls.d: Likewise.
	* gas/riscv/cv-mac-muls.s: Likewise.
	* gas/riscv/cv-mac-mulsn.d: Likewise.
	* gas/riscv/cv-mac-mulsn.s: Likewise.
	* gas/riscv/cv-mac-mulsrn.d: Likewise.
	* gas/riscv/cv-mac-mulsrn.s: Likewise.
	* gas/riscv/cv-mac-mulu.d: Likewise.
	* gas/riscv/cv-mac-mulu.s: Likewise.
	* gas/riscv/cv-mac-mulun.d: Likewise.
	* gas/riscv/cv-mac-mulun.s: Likewise.
	* gas/riscv/cv-mac-mulurn.d: Likewise.
	* gas/riscv/cv-mac-mulurn.s: Likewise.
	* gas/riscv/cv-mac-march-rv32i-xcorev.s: Likewise.
	* gas/riscv/cv-mac-march-rv32i-xcorev.d: Likewise.
	* gas/riscv/cv-hwloop-march-rv32i-xcorev.s: Renamed to seperate from
        mac.
	* gas/riscv/cv-hwloop-march-rv32i-xcorev.d: Likewise.

include/ChangeLog.COREV:

	* opcode/riscv-opc.h: Added CORE-V multiply accumulate support.
	* opcode/riscv.h: Likewise.

opcodes/ChangeLog.COREV:

	* riscv-dis.c: Added CORE-V multiply accumulate support.
	* riscv-opc.c: Likewise.

Signed-off-by: Jessica Mills <jessica.mills@embecosm.com>

### CODE REVIEW

File: gas/ChangeLog.COREV [hunk 1]
```diff
@@ -6,7 +10,7 @@
 	* config/tc-riscv.c (md_apply_fix): Renamed macros for clarity -
 	ENCODE_I1TYPE_UIMM -> ENCODE_CV_HWLP_UIMM5
 
-2020-10-05  Mary Bennett <mary.bennett@embecosm.com>
+2020-10-05  Mary Bennett  <mary.bennett@embecosm.com>
```
Comment (jeremybennett) [inline gas/ChangeLog.COREV:14]: Your merge went wrong - you've introduced a space before `<mary`

File: gas/config/tc-riscv.c [hunk 2]
```diff
@@ -988,12 +991,17 @@ validate_riscv_insn (const struct riscv_opcode *opc, int length)
       case 'b':
 	if (*p == '1')
 	  {
-	    used_bits |= ENCODE_ITYPE_IMM(-1U); /* For loop I type pc rel displacement */
+	    used_bits |= ENCODE_ITYPE_IMM(-1U);
```
Comment (jeremybennett) [inline gas/config/tc-riscv.c:994]: Was there a reason to get rid of the comment?

File: gas/config/tc-riscv.c [hunk 3]
```diff
@@ -988,12 +991,17 @@ validate_riscv_insn (const struct riscv_opcode *opc, int length)
...
 	    ++p; break;
 	  }
 	else if (*p == '2')
 	  {
-	    used_bits |= ENCODE_CV_HWLP_UIMM5(-1U); /* For loop I1 type pc rel displacement */
+	    used_bits |= ENCODE_CV_HWLP_UIMM5(-1U);
```
Comment (jeremybennett) [inline gas/config/tc-riscv.c:999]: Same here - why remove the comment?

File: gas/testsuite/gas/riscv/cv-hwloop-march-rv32i-xcorev.s [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# Testing xcorev march option works for all CORE-V extensions
+# xcorev march option works for all CORE-V hwloop extensions
 target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-hwloop-march-rv32i-xcorev.s:2]: Bad merge - I think the deleted comment is the correct one!

File: gas/testsuite/gas/riscv/cv-mac-mac.s [hunk 5]
```diff
@@ -0,0 +1,4 @@
+target:
+	cv.mac t0, t1, t2
+	cv.mac t4, t2, t0
+	cv.mac t3, t5, t1
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mac.s:4]: Needs a comment to explain what the test is

File: gas/testsuite/gas/riscv/cv-mac-machhsn.s [hunk 6]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-machhsn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-machhsrn.s [hunk 7]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-machhsrn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-machhun.s [hunk 8]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-machhun.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-machhurn.s [hunk 9]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-machhurn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-macsn.s [hunk 10]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-macsn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-macsrn.s [hunk 11]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-macsrn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-macun.s [hunk 12]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-macun.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-macurn.s [hunk 13]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-macurn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-msu.s [hunk 14]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-msu.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulhhs.s [hunk 15]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulhhs.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulhhsn.s [hunk 16]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulhhsn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulhhsrn.s [hunk 17]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulhhsrn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulhhu.s [hunk 18]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulhhu.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulhhun.s [hunk 19]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulhhun.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulhhurn.s [hunk 20]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulhhurn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-muls.s [hunk 21]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-muls.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulsn.s [hunk 22]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulsn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulsrn.s [hunk 23]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulsrn.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulu.s [hunk 24]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulu.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulun.s [hunk 25]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulun.s:1]: Needs a comment to explain the test.

File: gas/testsuite/gas/riscv/cv-mac-mulurn.s [hunk 26]
```diff
@@ -0,0 +1,4 @@
+target:
```
Comment (jeremybennett) [inline gas/testsuite/gas/riscv/cv-mac-mulurn.s:1]: Needs a comment to explain the test.

File: include/opcode/riscv-opc.h [hunk 27]
```diff
@@ -866,6 +866,41 @@
 #define MATCH_HWLP_COUNTI 0x0307b
 #define MATCH_HWLP_SETUP  0x0407b
 #define MATCH_HWLP_SETUPI 0x0507b
+/* Multiply accumulate */
```
Comment (jeremybennett) [inline include/opcode/riscv-opc.h:869]: Perhaps explicitly "CORE-V multiply accumulate"

File: opcodes/riscv-dis.c [hunk 28]
```diff
@@ -286,7 +286,12 @@ print_insn_args (const char *d, insn_t l, bfd_vma pc, disassemble_info *info)
 	      (*info->print_address_func) (info->target, info);
 	      break;
 	    }
-	/* Fall through.  */
+	  else if (d[1] == '3')
+	    {
+	      print (info->stream, "%d", ((int) EXTRACT_CV_MAC_UIMM5 (l)));
+	      ++d;
+	      break;
+	    }
```
Comment (jeremybennett) [inline opcodes/riscv-dis.c:294]: Still need the "Fall through" comment at the end of the case (after line 294)

File: opcodes/riscv-opc.c [hunk 29]
```diff
@@ -796,6 +796,34 @@ const struct riscv_opcode riscv_opcodes[] =
 {"cv.setup",  0, INSN_CLASS_COREV_HWLP, "di,s,b1",  	MATCH_HWLP_SETUP,  MASK_HWLP_SETUP, match_opcode, 0},
 {"cv.setupi", 0, INSN_CLASS_COREV_HWLP, "di,ji,b2", 	MATCH_HWLP_SETUPI, MASK_HWLP_SETUPI,match_opcode, 0},
 
+/* Multiply accumulate */
```
Comment (jeremybennett) [inline opcodes/riscv-opc.c:799]: Perhaps "CORE-V multiply accumulate"

**Review by jeremybennett — CHANGES_REQUESTED** [review]
Overall very good, but a number of small nits to fix. Some of the rebase on top of earlier work by @MaryBennett has gone wrong. In a number of places comments have been lost, which should probably be preserved. A number of tests lack comments to say what they do.

Once these are fixed, good to merge.

File: gas/testsuite/gas/riscv/cv-mac-mac.s (hunk 5 above)
Comment (jessicamills) [reply gas/testsuite/gas/riscv/cv-mac-mac.s:4]: @jeremybennett We don't comment on these tests as they are just a test for the instruction in the test name. We test three acceptable values. We haven't commented on similar tests in hwloop, so shall we go back and comment the hwloop ones also?

File: gas/ChangeLog.COREV (hunk 1 above)
Comment (jessicamills) [reply gas/ChangeLog.COREV:14]: @jeremybennett these should be double spaced, I just saw the error on Mary's previous changelog so corrected it

File: gas/config/tc-riscv.c (hunk 3 above)
Comment (jessicamills) [reply gas/config/tc-riscv.c:999]: @jeremybennett The comments relate to I1TYPE, I3TYPE etc. which made no sense. Mary renamed them in the previous commit but we forgot to take them out here

File: gas/testsuite/gas/riscv/cv-hwloop-march-rv32i-xcorev.s (hunk 4 above)
Comment (jessicamills) [reply gas/testsuite/gas/riscv/cv-hwloop-march-rv32i-xcorev.s:2]: @jeremybennett The message has changed as the test for the xcorev march option was split into two tests, one for each set of instructions i.e. hwlp & mac

File: include/opcode/riscv-opc.h (hunk 27 above)
Comment (jessicamills) [reply include/opcode/riscv-opc.h:869]: @jeremybennett this follows on from the headings:
/* CORE-V Specific Instructions  */
/* Hardware loops  */
Shall I go back and amend the hwlp heading also?

**Comment by simonpcook** [comment]
The commit message of this patch needs fixing, it has mixed whitespace in the ChangeLog text. ChangeLog itself seems to be fine, but haven't checked every line.

File: opcodes/riscv-dis.c (hunk 28 above)
Comment (jeremybennett) [reply opcodes/riscv-dis.c:294]: Actually this may be a bit more complex. Needs further investigation, but it certainly is not right as it stands

Comment (jeremybennett) [reply opcodes/riscv-dis.c:294]: Fallthrough comment now added

**Review by jeremybennett — APPROVED** [review]

**Comment by jeremybennett** [comment]
All approved by review. Good to merge.

**Pull request merged**